### ISSUE TITLE
Write a script to automatically propose PR's due to a new Kubernetes release

### ISSUE THREAD

**Issue by tstromberg** [kubernetes/minikube#4392]
I would love a script that we can run from a cronjob or Jenkins that monitors https://github.com/kubernetes/kubernetes/releases for new releases.

The script should propose a PR for any release newer than NewestKubernetesVersion:
 https://github.com/kubernetes/minikube/blob/f8fa0748c5a474bc2a62816ce318a6740fc00fea/pkg/minikube/constants/constants.go#L168 

If the release is a non-beta release, it should also change DefaultKubernetesVersion in the same PR:

https://github.com/kubernetes/minikube/blob/f8fa0748c5a474bc2a62816ce318a6740fc00fea/pkg/minikube/constants/constants.go#L165

This script should either build upon or replace  https://github.com/kubernetes/minikube/blob/master/hack/kubernetes_version/update_kubernetes_version.go - which makes the changes necessary in our code base, but is designed for interactive use and does not know how to look for new Kubernetes releases.

We can provide a github token via an environment variable.

**Comment by medyagh** [comment]
@tstromberg do we want the latest stable release ? or the latest tag of any kind (release, pre-release...) ?

**Comment by tstromberg** [comment]
`NewestKubernetesVersion` should be the most recent possible release: beta or RC is fine. 

`DefaultKubernetesVersion` should be the most recent official stable release.

**Comment by sjortiz** [comment]
Hey, I would like to work on this!

**Comment by tstromberg** [comment]
@sjortiz - If you are still interested, let me know how I can help.

**Comment by sjortiz** [comment]
hey, Sorry for the late response, I honestly forgot about this as it has been a couple of months since I first asked.

Yeah, I'd love to work on this!
@tstromberg

**Comment by sjortiz** [comment]
Could you provide me any insights on where to start?

**Comment by tstromberg** [comment]
@sjortiz - sure!

## Initial iteration

I would start by implementing a standalone program that polls GitHub for the latest Kubernetes release, and if differs the current value used by minikube, run or import `hacks/update_kubernetes_version`, and send a PR out. This program should be located in the `hacks/` subdirectory. Two reference programs that might help:

* https://github.com/kubernetes/minikube/blob/master/hack/release_notes/listpullreqs.go - shows how to query for releases, although it's querying for minikube releases.
* https://github.com/kubernetes/minikube/blob/master/hack/kubernetes_version/update_kubernetes_version.go - example script for updating `constants.DefaultKubernetesVersion`. I would execute it, or turn it into a library. The script may not work properly right now.

This first iteration should filter out -beta and -rc releases.

We will then take this script and set it to run on a periodic basis on our test servers.

## Further iteration

* script should query if there is already a PR with the same title, and if so, don't propose duplicates.
* -beta and -rc releases should update `constants.LatestKubernetesVersion`

## Documentation

https://godoc.org/github.com/google/go-github/github has the Go API documentation for the github library that we tend to use. 

Thank you for volunteering!

**Comment by tstromberg** [comment]
@sjortiz - Does this still issue seem reasonable to resolve?

**Comment by anorek** [comment]
@tstromberg regarding the filter of release version, is it enough to filter out pre-releases?

**Comment by gbraad** [comment]
@sjortiz any feedback on progress would be appreciated, else I would
suggest to drop the assignment and allow others to work on this.

**Comment by anorek** [comment]
@sjortiz @tstromberg could i take this issue?

**Comment by tstromberg** [comment]
@anorek - please do!

Filtering out pre-releases for the CurrentVersion sounds good.

**Comment by tstromberg** [comment]
Increasing the priority as we really would like to end the year with this roadmap item fixed.

**Comment by anorek** [comment]
will try :) might need some assist like the authentication key and so

**Comment by tstromberg** [comment]
@anorek - I'll be happy to help with any credential issues.

**Comment by medyagh** [comment]
@anorek are you still working this ?

**Comment by anorek** [comment]
@medyagh  been busy for a while but I have most of the code ready. Will draft pr within 2 weeks

**Comment by taciomcosta** [comment]
>     * ... hack/kubernetes_version/update_kubernetes_version.go - ... I think the script may  broken now that we've changed the testdata paths, ...
> 

Hi!
I'd like to send a PR to update the testdata file path mentioned by @tstromberg . May I link to this same issue?

**Comment by medyagh** [comment]
@anorek are you still woking on this?

**Comment by medyagh** [comment]
@anorek I haven't heard from you, are you still interested in finishing this?

**Comment by anorek** [comment]
@tstromberg is github action acceptable for this?

**Comment by sharifelgamal** [comment]
Github actions seem perfectly acceptable for this usecase.

**Comment by medyagh** [comment]
This should be added to one of our automation tasks in q4

**Comment by prezha** [comment]
@tstromberg, @medyagh, @sharifelgamal: i'd like to give it a try, if you don't mind :)

just three things to clarify/confirm:
a) re: `NewestKubernetesVersion` - should be the most recent possible release: beta or RC" - ie, _**not** alpha_ as well?
b) `testData` should use `DefaultKubernetesVersion`, `NewestKubernetesVersion`, or both (warning: in case it's 'both' - i'll have more questions)?
https://github.com/kubernetes/minikube/blob/f5b74f9eb41d459d8ed72ab3eb9434fae058bf73/hack/kubernetes_version/update_kubernetes_version.go#L80
c) script should update all respective `<major>.<minor>` versions in `testData` folder (in respective `default.yaml`), not just the latest one? 
https://github.com/kubernetes/minikube/blob/0a6b52d90de79179616f87b4df507a343e5ee53e/hack/kubernetes_version/update_kubernetes_version.go#L66
if you agree, i'd work on a PR for iteration 0: "implementing a standalone program that polls GitHub for the latest Kubernetes release", which would improve on the existing "interactive use and does not know how to look for new Kubernetes releases" (ie, next iteration would be checking and automatically creating a new PR)?

**Comment by tstromberg** [comment]
@prezha - Thanks for the questions!

a) I suggest ignoring alphas for NewestKubernetesVersion initially. We can always add them back in once we are confident in the workflow.
b) I don't think our testData should rely on either exact version at this point: but if so, let's stick with DefaultKubernetesVersion for now?
c) I think this part is mostly obsolete. We do need to still add testdata for major new versions, but let's save that for a subsequent PR.

The important thing is to iterate the version in constants.go and verify that `make test` passes. Automating the update  of  DefaultKubernetesVersion is more important than LatestKubernetesVersion (more toil & more impact). I suggest for the initial PR we focus on that.

**Comment by prezha** [comment]
thank you @tstromberg for your responses!

in summary:
a) ok, we'll use only the `-rc` or `-beta` for `NewestKubernetesVersion`, for now,
b) ok, let's stick with `DefaultKubernetesVersion` in `testData`, for now,
c) ok, we can cover the latest set of major.minor versions later (currently in `pkg/minikube/bootstrapper/bsutil/testdata` there are subfolders named `v1.11`..`1.19`)

conclusion: we'll focus on updating the version(s) in pkg/minikube/constants/constants.go and verify that `make test` passes

**Comment by prezha** [comment]
@tstromberg i've just created a pr for this first iteration - please have a look and let me know your thoughts; thanks!

**Comment by prezha** [comment]
thank you @medyagh for the review and approval/merge

shall i continue working on this according to the original scope - ie, automatically propose a pr for any release newer than current *KubernetesVersions values (ideally one/unique pr per each detected versions mismatch)?

if so: if the script would also run periodically and independently of (isolated from) the rest of the code - what would be the reference minikube release (and *KubernetesVersions) to check and send pr against: eg, latest stable minikube release or ?

**Comment by tstromberg** [comment]
@prezha - Sounds good. Here's how I see this working, please let me know if you have any alternative suggestions:

* We setup a cronjob that checks out the minikube source repository and runs this script. 
* This script is run, checking against the values in the local source tree
* If there is a new Kubernetes version, it updates the local source tree, and if there are no open PR's which have the new version in the title, it proposes a PR.

I have no idea how to suggest proposing a PR, unfortunately. I understand there is at least a GitHub command-line we could use now. Sound OK?

**Comment by prezha** [comment]
@tstromberg - sounds great and in line with my thoughts: follows the details, and i've created a [pr #9384](https://github.com/kubernetes/minikube/pull/9384) with changes in the script:

### notes:

- `UPDATE_TARGET` env var controlles what will be updated - options:
  = unset/absent, `fs`, or `all`: local filesystem repo will be updated - similarly to the old behaviour
  = `gh` or `all`: a pr against origin (`kubernetes/minikube`) github repo would be automatically created, but only if the same one does not exist already, and will 'mention' this issue here
- `GITHUB_TOKEN` is required for `gh` or `all` `UPDATE_TARGET`

- desired changes are defined as `plan`, which is flexible (ie, supports regex and go text templates) and placed at the top of the script - easy to amend and extend; parseability checks in place

- aob:
  = used `github.com/google/go-github/v32/github`: pure go code ;) - no need to call `git` or `gh` cli from the script
  = used `k8s.io/klog/v2` instead of `glog`
  = `go.mod` & `go.sum` amended accordingly (for klog/v2 and github/v32)
  = cron job cmd could be something like: `rm -rf ~/minikube && git clone https://github.com/kubernetes/minikube.git ~/minikube && go run ~/minikube/hack/kubernetes_version/update_kubernetes_version.go`

- all `make test` passed

- testing & example output:
## for unset `UPDATE_TARGET`:

### 1st run:

```
I1003 00:52:44.696940    1777 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 00:52:44.697055    1777 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 00:52:44.697581    1777 update_kubernetes_version.go:172] Local repo updated
```

### git diff:

```
diff --git a/pkg/minikube/constants/constants.go b/pkg/minikube/constants/constants.go
index fd330a96d..a533e0e09 100644
--- a/pkg/minikube/constants/constants.go
+++ b/pkg/minikube/constants/constants.go
@@ -27,10 +27,10 @@ import (

 const (
        // DefaultKubernetesVersion is the default Kubernetes version
-       DefaultKubernetesVersion = "v1.19.2"
+       DefaultKubernetesVersion = "v1.23.456"
        // NewestKubernetesVersion is the newest Kubernetes version to test against
        // NOTE: You may need to update coreDNS & etcd versions in pkg/minikube/bootstrapper/images/images.go
-       NewestKubernetesVersion = "v1.19.2"
+       NewestKubernetesVersion = "v2.34.457-rc.89"
        // OldestKubernetesVersion is the oldest Kubernetes version to test against
        OldestKubernetesVersion = "v1.13.0"
        // DefaultClusterName is the default nane for the k8s cluster
diff --git a/site/content/en/docs/commands/start.md b/site/content/en/docs/commands/start.md
index 87538fb59..b1f036128 100644
--- a/site/content/en/docs/commands/start.md
+++ b/site/content/en/docs/commands/start.md
@@ -67,7 +67,7 @@ minikube start [flags]
       --interactive                       Allow user prompts for more information (default true)
       --iso-url strings                   Locations to fetch the minikube ISO from. (default [https://storage.googleapis.com/minikube/iso/minikube-v1.13.1.iso,https://github.com/kubernetes/minikube/releases/download/v1.13.1/minikube-v1.13.1.iso,https://kubernetes.oss-cn-hangzhou.aliyuncs.com/minikube/iso/minikube-v1.13.1.iso])
       --keep-context                      This will keep the existing kubectl context and will create a minikube context.
-      --kubernetes-version string         The Kubernetes version that the minikube VM will use (ex: v1.2.3, 'stable' for v1.19.2, 'latest' for v1.19.2). Defaults to 'stable'.
+      --kubernetes-version string         The Kubernetes version that the minikube VM will use (ex: v1.2.3, 'stable' for v1.23.456, 'latest' for v2.34.457-rc.89). Defaults to 'stable'.
       --kvm-gpu                           Enable experimental NVIDIA GPU support in minikube
       --kvm-hidden                        Hide the hypervisor signature from the guest in minikube (kvm2 driver only)
       --kvm-network string                The KVM network name. (kvm2 driver only) (default "default")

```

### 2nd run:

```
I1003 00:53:30.207909    3317 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 00:53:30.208360    3317 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 00:53:30.209539    3317 update_kubernetes_version.go:170] Local repo update skipped: nothing changed
```

## for `UPDATE_TARGET` set to `gh` and w/ `GITHUB_TOKEN` set:

### 1st run:

```
I1003 01:06:26.158314   21354 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 01:06:26.158472   21354 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 01:06:34.204164   21354 update_kubernetes_version.go:289] PR commit '0d49507f0f4d6fb8c6f8cf646a65a737371da2e8' created: https://github.com/prezha/minikube/commit/0d49507f0f4d6fb8c6f8cf646a65a737371da2e8
I1003 01:06:34.921082   21354 update_kubernetes_version.go:303] PR branch 'update-kubernetes-version_0d49507' created: https://api.github.com/repos/prezha/minikube/git/refs/heads/update-kubernetes-version_0d49507
I1003 01:06:36.615138   21354 update_kubernetes_version.go:199] PR created: https://github.com/triplepoint-tech/minikube/pull/3
```

### 2nd run:

```
I1003 01:06:52.070881   22035 update_kubernetes_version.go:161] Kubernetes versions: 'stable' is v1.23.456 and 'latest' is v2.34.457-rc.89
I1003 01:06:52.071322   22035 update_kubernetes_version.go:163] The Plan:
{
  "pkg/minikube/constants/constants.go": {
    "replace": {
      "DefaultKubernetesVersion = \\\".*": "DefaultKubernetesVersion = \"v1.23.456\"",
      "NewestKubernetesVersion = \\\".*": "NewestKubernetesVersion = \"v2.34.457-rc.89\""
    }
  },
  "site/content/en/docs/commands/start.md": {
    "replace": {
      "'latest' for .*\\)": "'latest' for v2.34.457-rc.89)",
      "'stable' for .*,": "'stable' for v1.23.456,"
    }
  }
}
I1003 01:06:52.434969   22035 update_kubernetes_version.go:190] PR create skipped: already exists (https://github.com/triplepoint-tech/minikube/pull/3)
```

please share your thoughts

thanks!

**Comment by prezha** [comment]
@tstromberg now that pr #9384 (related to this issue) is approved, i would propose to reuse the same approach also for issues eg, #9264 [`update go version to be 1.15 in go mod`], #9420 [`automate pushing kicbase image from snapshot to stable after release`] (and potentially others) that have similar requirements (ie, fully/semi-automated patching of specific files + additional actions)

the idea is that we just need to define what we want to update and with which values, and leave implementation details to available functions

i have restructured the code accordingly and also expanded to include functionalities for above-mentioned issues

so, currently, with this, we can: track new kubernetes stable & latest versions, new goland stable versions, update any local fs files, update any files on github, automatically create pr for those changes (if one does not exists already), update all three remote container registries we use here (ie, gcr, docker, and github)

i've created a new pr #9464 here with these changes and additions (it would create a separate `update` folder under `/hack`) for you review - please have a look and share your thoughts

thanks!

**Comment by prezha** [comment]
btw, Kubernetes released v1.19.3 yesterday, so maybe we could give the script a go?
